ds3-autogen-c/src/main/java/com/spectralogic/ds3autogen/c/CCodeGenerator.java

```diff
@@ -59,121 +63,220 @@ public CCodeGenerator() {
         config.setClassForTemplateLoading(CCodeGenerator.class, "/templates");
     }
 
-    public Path getRequestOutputPath(final Request request) {
-        return Paths.get(outputDirectory + "/ds3_c_sdk/src/requests/" + RequestHelper.getNameRootUnderscores(request.getName()) + ".c");
+    public void setFileUtils(final FileUtils fileUtils) {
+        this.fileUtils = fileUtils;
     }
 
-    public Path getTypeOutputPath(final Type type) {
-        return Paths.get(outputDirectory + "/ds3_c_sdk/src/types/" + TypeHelper.getNameUnderscores(type.getName()) + ".h");
+    public void setSpec(final Ds3ApiSpec spec) {
+        this.spec = spec;
     }
 
     @Override
     public void generate(final Ds3ApiSpec spec, final FileUtils fileUtils, final Path destDir) throws IOException {
         this.spec = spec;
         this.fileUtils = fileUtils;
-        this.outputDirectory = destDir;
 
         try {
-            generateCommands();
+            generateDs3_H();
+            generateDs3_C();
         } catch (final TemplateException e) {
             e.printStackTrace();
+        } catch (final ParseException e) {
+            e.printStackTrace();
         }
     }
 
-    private void generateCommands() throws IOException, TemplateException {
+    public void generateDs3_H() throws IOException, ParseException {
+        final Path path = Paths.get("src/ds3.h");
+        final OutputStream outputStream = fileUtils.getOutputFile(path);
+
+        // Enums
+        generateEnums(outputStream);
+
+        // ResponseStructs
+        generateTypedefStructs(outputStream);
+
+        // InitFunctionPrototypes
+        generateInitRequestPrototypes(outputStream);
+
+        // request prototypes
+        generateRequestPrototypes(outputStream);
+
+        // Free ResponseStruct prototypes
+        generateFreeResponseStructPrototypes(outputStream);
+    }
+
+    public void generateEnums(final OutputStream outputStream) throws IOException {
         if (ConverterUtil.hasContent(spec.getTypes())) {
```

Comment:
Can remove a layer of nesting if you reverse this if statement to: 
if (isEmpty(spec.getTypes())) { return; }